src/FsAutoComplete/FsAutoComplete.Lsp.fs

```
@@ -1188,13 +1188,60 @@ type FSharpLspServer(backgroundServiceEnabled: bool, state: State, lspClient: FS
         let fileName = doc.GetFilePath() |> Utils.normalizePath
         let! res = commands.FormatDocument fileName
         match res with
-        | Ok (lines, formatted) ->
+        | Ok (FormatDocumentResponse.Formatted (lines, formatted)) ->
             let range =
                 let zero = { Line = 0; Character = 0 }
                 let lastPos = lines.GetLastFilePosition()
                 { Start = zero; End = fcsPosToLsp lastPos }
 
             return LspResult.success(Some([| { Range = range; NewText = formatted  } |]))
+        | Ok (FormatDocumentResponse.UnChanged | FormatDocumentResponse.Ignored) ->
+          return LspResult.success None
+        | Ok FormatDocumentResponse.ToolNotPresent ->
+              let actions =
+                [| if Option.isSome rootPath then { Title = "Install locally" }
+                   { Title = "Install globally" } |]
+              let! response =
+                lspClient.WindowShowMessageRequest { Type = MessageType.Warning
+                                                     Message = "No Fantomas install was found."
+                                                     Actions = Some actions }
+
+              match response with
+              | Ok (Some { Title = "Install locally" }) ->
+                let didInstall =
+                  rootPath
+                  |> Option.map (fun rootPath->
+                    let dotConfig = Path.Combine(rootPath, ".config", "dotnet-tools.json")
+                    if not (File.Exists dotConfig) then
+                      Process.Start("dotnet", "new tool-manifest").WaitForExit(5000) |> ignore
+                    else
+                      let dotConfigContent = File.ReadAllText dotConfig
+                      if dotConfigContent.Contains("fantomas-tool") then
+                        // uninstall a older, non-compatible version of fantomas-tool
+                        Process.Start("dotnet", @"tool uninstall fantomas-tool").WaitForExit(5000) |> ignore
```

Any objection to bringing in https://github.com/Tyrrrz/CliWrap?
As `Process.Start `can now fail silently.

---

Nope, go on ahead!